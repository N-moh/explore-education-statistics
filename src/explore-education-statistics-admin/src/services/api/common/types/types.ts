import { User } from '../../../PrototypeLoginService';

export interface IdLabelPair {
  id: string;
  label: string;
}

export interface UserContact {
  name: string;
  email: string;
  telNo: string;
}

export interface Topic {
  id: string;
  title: string;
  theme: IdLabelPair;
}

export interface TimePeriodCoverage {
  label: string;
  code: string;
  startDate: Date;
}

export enum ApprovalStatus {
  Approved,
  ReadyToReview,
}

export interface ReleaseStatus {
  approvalStatus: ApprovalStatus;
  isNew: boolean;
  isLive: boolean;
  isLatest: boolean;
  lastEdited: Date;
  lastEditor: User;
  published: Date;
  nextRelease: Date;
}

export interface ReleaseDataType {
  id: string;
  title: string;
}

export interface Comment {
  id: string;
  author: User;
  datetime: Date;
  content: string;
}

export interface Release {
  id: string;
  releaseName: string;
  timePeriodCoverage: TimePeriodCoverage;
  scheduledPublishDate: DayMonthYearValues;
  nextReleaseExpectedDate: Date;
  releaseType: IdLabelPair;
  slug: string;
  status: ReleaseStatus;
  lead: UserContact;
  dataType: ReleaseDataType;
  comments: Comment[];
}

export interface LegacyRelease {
  id: string;
  description: string;
  url: string;
}

export interface Publication {
  id: string;
  slug: string;
  title: string;
  description: string;
  dataSource: string;
  summary: string;
  nextUpdate: string;
  releases: Release[];
  legacyReleases: LegacyRelease[];
  topic: Topic;
  contact: UserContact;
  methodology: IdLabelPair;
  owner: User;
}

export interface ReleaseSetupDetails {
  id: string;
  publicationTitle: string;
  timePeriodCoverageCode: string;
  timePeriodCoverageStartDate: Date;
  releaseType: IdLabelPair;
  leadStatisticianName: string;
  scheduledPublishDate: DayMonthYearValues;
  nextReleaseExpectedDate?: Date;
}

export interface DayMonthYearValues {
  day?: number;
  month?: number;
  year?: number;
}

export const dateToDayMonthYear = (date?: Date) => {
  return {
    day: date && date.getDate(),
    month: date && date.getMonth() + 1,
    year: date && date.getFullYear(),
  };
};

export const dayMonthYearToDate = (dmy: DayMonthYearValues) => {
  if (!(dmy.day && dmy.month && dmy.year)) {
    throw Error(
      `Couldn't convert DayMonthYearValues ${JSON.stringify(
        dmy,
      )} to Date - missing required value`,
    );
  }
  return new Date(dmy.year, dmy.month - 1, dmy.day);
};
