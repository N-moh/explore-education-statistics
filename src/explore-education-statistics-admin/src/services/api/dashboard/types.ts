import { User } from '@admin/services/PrototypeLoginService';
import {
  DayMonthYearValues,
  IdLabelPair,
  UserContact,
} from '@admin/services/api/common/types/types';

export enum ReleaseApprovalStatus {
  Approved,
  ReadyToReview,
}

export interface ThemeAndTopics {
  title: string;
  id: string;
  topics: {
    title: string;
    id: string;
  }[];
}

export interface AdminDashboardRelease {
  id: string;
  status: ReleaseApprovalStatus;
  latestRelease: boolean;
  live: boolean;
  dateRangeLabel: string;
  timePeriodCoverage: IdLabelPair;
  leadStatistician: UserContact;
  lastEditedUser: User;
  lastEditedDateTime: string;
  scheduledPublishDate: DayMonthYearValues;
  nextReleaseExpectedDate: DayMonthYearValues;
}

export interface AdminDashboardPublication {
  id: string;
  title: string;
  methodology: IdLabelPair;
  releases: AdminDashboardRelease[];
}
