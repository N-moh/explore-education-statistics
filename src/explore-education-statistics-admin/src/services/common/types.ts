export interface IdTitlePair {
  id: string;
  title: string;
}

export interface IdLabelPair {
  id: string;
  label: string;
}

export interface ValueLabelPair {
  value: string;
  label: string;
}

export interface ContactDetails {
  id: string;
  contactName: string;
  contactTelNo: string;
  teamEmail: string;
  teamName?: string;
}

export interface UserDetails {
  id: string;
  name: string;
}

export interface TimePeriodCoverageGroup {
  category: {
    label: string;
  },
  timeIdentifiers: {
    identifier: ValueLabelPair;
  }[];
}

export interface BasicPublicationDetails {
  id: string;
  title: string;
  contact?: ContactDetails;
}

export interface DayMonthYearValues {
  day?: number;
  month?: number;
  year?: number;
}

export interface DayMonthYearInputs {
  day: string;
  month: string;
  year: string;
}

export const dayMonthYearValuesToInputs = (
  dmy: DayMonthYearValues,
): DayMonthYearInputs => ({
  day: dmy.day ? dmy.day.toString() : '',
  month: dmy.month ? dmy.month.toString() : '',
  year: dmy.year ? dmy.year.toString() : '',
});

export const dayMonthYearInputsToValues = (
  dmy: DayMonthYearInputs,
): DayMonthYearValues => ({
  day: dmy.day ? parseInt(dmy.day, 10) : undefined,
  month: dmy.month ? parseInt(dmy.month, 10) : undefined,
  year: dmy.year ? parseInt(dmy.year, 10) : undefined,
});

export const dateToDayMonthYear = (date?: Date) => {
  return {
    day: date && date.getDate(),
    month: date && date.getMonth() + 1,
    year: date && date.getFullYear(),
  };
};

export const emptyDayMonthYear = (): DayMonthYearInputs => ({
  day: '',
  month: '',
  year: '',
});

export const dayMonthYearIsComplete = (dmy?: DayMonthYearValues) => {
  return dmy && dmy.day && dmy.month && dmy.year;
};

export const dayMonthYearIsEmpty = (dmy?: DayMonthYearValues) => {
  return !dmy || (!dmy.day && !dmy.month && !dmy.year);
};

export const dayMonthYearToDate = (dmy: DayMonthYearValues) => {
  if (!dayMonthYearIsComplete(dmy)) {
    throw Error(
      `Couldn't convert DayMonthYearValues ${JSON.stringify(
        dmy,
      )} to Date - missing required value`,
    );
  }
  return new Date(dmy.year || 0, (dmy.month || 0) - 1, dmy.day);
};

export const dayMonthYearInputsToDate = (dmy: DayMonthYearInputs): Date =>
  dayMonthYearToDate(dayMonthYearInputsToValues(dmy));
