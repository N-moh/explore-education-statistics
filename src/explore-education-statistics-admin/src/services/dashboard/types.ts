import {
  ContactDetails,
  DayMonthYearValues,
  IdLabelPair,
  IdTitlePair,
  UserDetails,
} from '@admin/services/common/types';

export enum ReleaseApprovalStatus {
  Approved,
  ReadyToReview,
  None,
}

export interface ThemeAndTopics {
  title: string;
  id: string;
  topics: {
    title: string;
    id: string;
  }[];
}

export interface AdminDashboardRelease {
  id: string;
  status: ReleaseApprovalStatus;
  latestRelease: boolean;
  live: boolean;
  releaseName: string;
  timePeriodCoverage: IdLabelPair;
  contact: ContactDetails;
  lastEditedUser: UserDetails;
  lastEditedDateTime: string;
  publishScheduled: Date;
  published?: string;
  nextReleaseDate: DayMonthYearValues;
}

export interface AdminDashboardPublication {
  id: string;
  title: string;
  methodology?: IdTitlePair;
  releases: AdminDashboardRelease[];
  contact: ContactDetails;
}
