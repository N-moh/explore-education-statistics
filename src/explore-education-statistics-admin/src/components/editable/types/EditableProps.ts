export default interface EditableProps {
  canDelete?: boolean;
  editable?: boolean;
  onDelete?: () => void;
}
