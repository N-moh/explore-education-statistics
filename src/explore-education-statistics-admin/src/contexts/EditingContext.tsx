import React, {
  createContext,
  ReactNode,
  useContext,
  useMemo,
  useState,
} from 'react';
import noop from 'lodash/noop';

export interface UnSavedEdit {
  sectionId: string;
  blockIds: string[];
}

export interface EditingContextState {
  isEditing: boolean;
  setEditing: (isEditing: boolean) => void;
  isTablePreview: boolean;
  setTablePreview: (isTablePreview: boolean) => void;
  unSavedEdits: UnSavedEdit[];
  setUnSavedEdits: (unSavedEdits: UnSavedEdit[]) => void;
}

export const EditingContext = createContext<EditingContextState>({
  isEditing: false,
  setEditing: noop,
  isTablePreview: false,
  setTablePreview: noop,
  unSavedEdits: [],
  setUnSavedEdits: noop,
});

export function useEditingContext() {
  return useContext(EditingContext);
}

export interface EditingContextProviderProps {
  children: ReactNode | ((state: EditingContextState) => ReactNode);
  value?: {
    isEditing: boolean;
    isTablePreview?: boolean;
    unSavedEdits?: UnSavedEdit[];
  };
}

export const EditingContextProvider = ({
  children,
  value = {
    isEditing: true,
    isTablePreview: false,
  },
}: EditingContextProviderProps) => {
  const [isEditing, setEditing] = useState<boolean>(value.isEditing);
  const [isTablePreview, setTablePreview] = useState<boolean>(false);
  const [unSavedEdits, setUnSavedEdits] = useState<UnSavedEdit[]>([]);

  const state = useMemo<EditingContextState>(() => {
    return {
      isEditing,
      setEditing,
      isTablePreview,
      setTablePreview,
      unSavedEdits,
      setUnSavedEdits,
    };
  }, [isEditing, isTablePreview, unSavedEdits]);

  return (
    <EditingContext.Provider value={state}>
      {typeof children === 'function' ? children(state) : children}
    </EditingContext.Provider>
  );
};
