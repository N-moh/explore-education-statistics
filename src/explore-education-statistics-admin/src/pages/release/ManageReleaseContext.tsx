import { BasicPublicationDetails } from '@admin/services/common/types';
import * as React from 'react';

export interface ManageRelease {
  publication: BasicPublicationDetails;
  releaseId: string;
}

export default React.createContext<ManageRelease | undefined>(undefined);
