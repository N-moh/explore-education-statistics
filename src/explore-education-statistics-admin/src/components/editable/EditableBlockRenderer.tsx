import EditableDataBlock from '@admin/components/editable/EditableDataBlock';
import EditableHtmlRenderer from '@admin/components/editable/EditableHtmlRenderer';
import EditableMarkdownRenderer from '@admin/components/editable/EditableMarkdownRenderer';
import EditableProps from '@admin/components/editable/types/EditableProps';
import { useManageReleaseContext } from '@admin/pages/release/ManageReleaseContext';
import { EditableBlock } from '@admin/services/publicationService';
import { OmitStrict } from '@common/types';
import React, { useCallback } from 'react';

interface Props extends OmitStrict<EditableProps, 'onDelete'> {
  allowHeadings?: boolean;
  block: EditableBlock;
  onSave: (content: string) => void;
  onDelete: (blockId: string) => void;
}

function EditableBlockRenderer({
  block,
  editable,
  allowHeadings,
  onSave,
  canDelete = false,
  onDelete,
}: Props) {
  const { releaseId } = useManageReleaseContext();
  const id = `editableDataBlockRenderer-dataBlock-${block.id}`;

  const handleDelete = useCallback(() => {
    if (onDelete) {
      onDelete(block.id);
    }
  }, [block.id, onDelete]);

  switch (block.type) {
    case 'MarkDownBlock':
      return (
        <EditableMarkdownRenderer
          id={id}
          label="Block content"
          editable={editable}
          allowHeadings={allowHeadings}
          source={block.body}
          canDelete={canDelete}
          onDelete={handleDelete}
          onSave={onSave}
        />
      );
    case 'DataBlock':
      return (
        <div className="dfe-content-overflow">
          <EditableDataBlock
            id={id}
            dataBlock={block}
            editable={editable}
            releaseId={releaseId}
            canDelete={canDelete}
            onDelete={handleDelete}
          />
        </div>
      );
    case 'HtmlBlock':
      return (
        <EditableHtmlRenderer
          id={id}
          label="Block content"
          editable={editable}
          allowHeadings={allowHeadings}
          source={block.body}
          canDelete={canDelete}
          onDelete={handleDelete}
          onSave={onSave}
        />
      );
    default:
      return <div>Unable to edit content</div>;
  }
}

export default EditableBlockRenderer;
