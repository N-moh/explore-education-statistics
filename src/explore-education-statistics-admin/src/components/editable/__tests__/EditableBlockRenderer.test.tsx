import EditableBlockRenderer from '@admin/components/editable/EditableBlockRenderer';
import { EditingContext as EditingContext1 } from '@admin/contexts/EditingContext';
import ManageReleaseContext from '@admin/pages/release/ManageReleaseContext';
import { BasicPublicationDetails } from '@admin/services/common/types';
import { render } from '@testing-library/react';
import React from 'react';

describe('EditableBlockRenderer', () => {
  test('renders non-editable Markdown block correctly', () => {
    const { container } = render(
      <EditingContext1
        value={{
          isEditing: true,
        }}
      >
        <ManageReleaseContext.Provider
          value={{
            releaseId: '1',
            publication: {} as BasicPublicationDetails,
          }}
        >
          <EditableBlockRenderer
            block={{
              id: 'block-000',
              order: 0,
              comments: [],
              type: 'MarkDownBlock',
              body: 'test',
            }}
            onDelete={() => {}}
            onContentChange={async () => {}}
          />
        </ManageReleaseContext.Provider>
      </EditingContext1>,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders editable Markdown block correctly', () => {
    const { container } = render(
      <EditingContext1
        value={{
          isEditing: true,
        }}
      >
        <ManageReleaseContext.Provider
          value={{
            releaseId: '1',
            publication: {} as BasicPublicationDetails,
          }}
        >
          <EditableBlockRenderer
            canDelete
            editable
            block={{
              id: 'block-000',
              order: 0,
              comments: [],
              type: 'MarkDownBlock',
              body: 'test',
            }}
            onDelete={() => {}}
            onContentChange={async () => {}}
          />
        </ManageReleaseContext.Provider>
      </EditingContext1>,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });
});
