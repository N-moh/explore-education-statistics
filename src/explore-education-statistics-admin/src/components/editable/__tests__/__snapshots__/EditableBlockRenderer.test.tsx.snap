// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableBlockRenderer MarkDownBlock renders editable version correctly 1`] = `
<article>
  <div class="block">
    <div class="preview readOnly">
      <div class="editButton"
           role="button"
           tabindex="0"
      >
        <div>
          <p>
            test
          </p>
        </div>
      </div>
    </div>
  </div>
  <div class="group buttons">
    <button aria-disabled="false"
            class="govuk-button govuk-button--secondary"
            type="button"
    >
      Edit block
    </button>
    <button aria-disabled="false"
            class="govuk-button govuk-button--warning"
            type="button"
    >
      Remove block
    </button>
  </div>
</article>
`;

exports[`EditableBlockRenderer MarkDownBlock renders non-editable version correctly 1`] = `
<article>
  <div class="block">
    <div class="preview readOnly">
      <div>
        <p>
          test
        </p>
      </div>
    </div>
  </div>
  <div class="group buttons">
  </div>
</article>
`;
